### PR TITLE
properly merge AD values when no PLs

### PR DESCRIPTION
**Pull request by RoriCremer**

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/ReferenceConfidenceVariantContextMerger.java [hunk 1]
```diff
@@ -545,14 +545,18 @@ private GenotypesContext mergeRefConfidenceGenotypes(final VariantContext vc,
...
-                    genotypeBuilder.PL(PLs).AD(AD);
-                //clean up low confidence hom refs for better annotations later
+                    genotypeBuilder.PL(PLs);
+                // clean up low confidence hom refs for better annotations later
                 } else if (GenotypeGVCFsEngine.excludeFromAnnotations(g)) {
                     genotypeBuilder.alleles(Collections.nCopies(ploidy, Allele.NO_CALL));
                 }
+                if (g.hasAD()) {
+                    int[]  perSampleIndexesOfRelevantAlleles = AlleleSubsettingUtils.getIndexesOfRelevantAllelesForGVCF(remappedAlleles, targetAlleles, vc.getStart(), g, false);
+                    final int[] AD = g.hasAD() ? AlleleSubsettingUtils.generateAD(g.getAD(), perSampleIndexesOfRelevantAlleles) : null;
```
Comment (mcovarr) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/ReferenceConfidenceVariantContextMerger.java:555]: line 553 already tested for `g.hasAd()`
```suggestion
                    final int[] AD = AlleleSubsettingUtils.generateAD(g.getAD(), perSampleIndexesOfRelevantAlleles);
```

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/ReferenceConfidenceVariantContextMerger.java [hunk 2]
```diff
@@ -545,14 +545,18 @@ private GenotypesContext mergeRefConfidenceGenotypes(final VariantContext vc,
...
                     final int[] PLs = generatePL(g, genotypeIndexMapByPloidy);
-                    final int[] AD = g.hasAD() ? AlleleSubsettingUtils.generateAD(g.getAD(), perSampleIndexesOfRelevantAlleles) : null;
-                    genotypeBuilder.PL(PLs).AD(AD);
-                //clean up low confidence hom refs for better annotations later
+                    genotypeBuilder.PL(PLs);
+                // clean up low confidence hom refs for better annotations later
                 } else if (GenotypeGVCFsEngine.excludeFromAnnotations(g)) {
                     genotypeBuilder.alleles(Collections.nCopies(ploidy, Allele.NO_CALL));
                 }
+                if (g.hasAD()) {
```
Comment (kcibul) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/ReferenceConfidenceVariantContextMerger.java:553]: I would move this up instead so the code is less branched/duplicative… and you've also lost the else-if case when we have ADs.  like

```
if (g.hasPL() || g.hasAD()) {
  # do common stuff (like perSampleIndexesOfRelevantAlleles)
  if (g.hasPL() {
     # do PL only stuff
   }
   if (g.hasAD()) {   
      # do AD stuff
   }
} else if (GenotypeGVCFsEngine.excludeFromAnnotations(g)) {
   # as is
}

**Review by kcibul — CHANGES_REQUESTED** [review]
How about a unit test as well?  Likely you can add/modify an existing.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/ReferenceConfidenceVariantContextMerger.java [hunk 3]
```diff
@@ -545,14 +545,18 @@ private GenotypesContext mergeRefConfidenceGenotypes(final VariantContext vc,
...
+                    int[]  perSampleIndexesOfRelevantAlleles = AlleleSubsettingUtils.getIndexesOfRelevantAllelesForGVCF(remappedAlleles, targetAlleles, vc.getStart(), g, false);
```
Comment (mcovarr) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/ReferenceConfidenceVariantContextMerger.java:554]: I have no idea how expensive this operation is; might it be worth popping this a scoping level to be right under line 540 with a `if (g.hasPL() || g.hasAD()) {...` so it never gets called twice (lines 543 and 554)?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/ReferenceConfidenceVariantContextMerger.java (hunk 2 above)
Comment (RoriCremer) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/ReferenceConfidenceVariantContextMerger.java:553]: Do we want the GenotypeGVCFsEngine.excludeFromAnnotations(g) if there is no AD? I didn't think so, which I why I set it up the way I did

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/ReferenceConfidenceVariantContextMerger.java [hunk 4]
```diff
@@ -538,21 +538,26 @@ private GenotypesContext mergeRefConfidenceGenotypes(final VariantContext vc,
             final int ploidy = g.getPloidy();
             final GenotypeBuilder genotypeBuilder = new GenotypeBuilder(g);
             if (!doSomaticMerge) {
-                if (g.hasPL()) {
-                    // lazy initialization of the genotype index map by ploidy.
+                if (g.hasPL() | g.hasAD()) {
```
Comment (kcibul) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/ReferenceConfidenceVariantContextMerger.java:541]: Although technically the bitwise or operator `|` does what you want here, convention is to use conditional-or `||` unless you really want the non-conditional behavior.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/ReferenceConfidenceVariantContextMergerUnitTest.java [hunk 5]
```diff
@@ -205,6 +211,13 @@ public Object[][] makeReferenceConfidenceMergeData() {
                         new GenotypeBuilder("A_C_G.test2").PL(new int[]{40, 20, 30, 20, 10, 30}).alleles(noCalls).make(),
                         new GenotypeBuilder("A_C_G.test").PL(new int[]{40, 20, 30, 20, 10, 30}).alleles(noCalls).make()).make()});
 
+        // test creation of AD with proper allele indexing without PLs
+        // this needs to be multi-allelic
+        tests.add(new Object[]{"test12",Arrays.asList(vcA_C_G_noPLs, vcA_C_G_ALT_noPLs), loc, false, true,
+                new VariantContextBuilder(VCbase).alleles(A_C_G).genotypes(
+                        new GenotypeBuilder("A_C_G.test2").AD(new int[]{60,9,0}).alleles(noCalls).make(),
+                        new GenotypeBuilder("A_C_G.test").AD(new int[]{60,9,0}).alleles(noCalls).make()).make()});
```
Comment (ldgauthier) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/ReferenceConfidenceVariantContextMergerUnitTest.java:219]: The variant context has Arrays.asList(Aref, C, G, Allele.NON_REF_ALLELE) so the ADs should be of length 4.  Do you mean LAD?  If this is how you want it for GVS, then I'd like to make that more obvious some how.

Comment (RoriCremer) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/ReferenceConfidenceVariantContextMergerUnitTest.java:219]: hmmmm then I'm not testing what I think I'm testing....

Comment (RoriCremer) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/ReferenceConfidenceVariantContextMergerUnitTest.java:237]: in the test (on line 51) we hardcode the filtering out of the Allele.NON_REF_ALLELE
(  @param removeNonRefSymbolicAllele if true, remove the <NON_REF> allele from the merged VC ) so it's not local AD and should not have that additional AD value.

**Comment by RoriCremer** [comment]
merging w blessing from Jonn

**Pull request merged**